ignite/config/plugins/config.go

```diff
@@ -57,30 +35,29 @@ type Plugin struct {
 	// path: github.com/foo/bar/plugin1@v42
 	Path string `yaml:"path"`
 	// With holds arguments passed to the plugin interface
-	With map[string]string `yaml:"with"`
+	With map[string]string `yaml:"with,omitempty"`
 }
 
-// Clone returns an identical copy of the instance
-func (c *Config) Clone() (*Config, error) {
-	copy := Config{}
-	if err := mergo.Merge(&copy, c, mergo.WithAppendSlice); err != nil {
-		return nil, err
-	}
-
-	return &copy, nil
+// Path return the path of the config file.
+func (c Config) Path() string {
+	return c.path
 }
 
-// Decode decodes the config file values from YAML.
-func (c *Config) Decode(r io.Reader) error {
-	// if the error is end of file meaning an empty file on read return nil
-	if err := yaml.NewDecoder(r).Decode(c); err != io.EOF {
-		return err
+// Save persists a config yaml to a specified path on disk must be writable
+func (c *Config) Save() error {
+	errf := func(err error) error {
+		return fmt.Errorf("plugin config save: %w", err)
+	}
+	if c.path == "" {
+		return errf(errors.New("empty path"))
+	}
+	file, err := os.Create(c.path)
+	if err != nil {
+		return errf(err)
+	}
+	defer file.Close()
+	if err := yaml.NewEncoder(file).Encode(c); err != nil {
+		return errf(err)
```

Comment:
Maybe

```go
var ErrEmptyPath = errors.New("config path is empty")

func newConfigSaveErr(reason any) error {
	return fmt.Errorf("plugin config save failed: %w", err)
}

func (c *Config) Save() error {
	if c.path == "" {
		return newConfigSaveErr(ErrEmptyPath)
	}
	file, err := os.Create(c.path)
	if err != nil {
		return newConfigSaveErr(err)
	}
	defer file.Close()
	if err := yaml.NewEncoder(file).Encode(c); err != nil {
		return newConfigSaveErr(err)
```

Something I'm not sure about is if the case where the config path is empty could actually happen or if it would be a critical error that should panic.

Reply:
This `errf` pattern is just to avoid repetition of the error prefix `plugin config save failed: `. I don't think it's better to have it separate from the function scope, that's why I prefer to inline it inside the function.

About `ErrEmptyPath`, as you asked, the case *should never happen* because the user should always get a `Config` from `ParseDir`, and not create a `Config` manually. For that reason there's also no need for a custom error IMO.